File: Zoo_Java/src/Zoo/Animals/Antelope.java
package Zoo.Animals;

import java.awt.Color;
import java.util.Random;

import Zoo.Animal;
import Zoo.Organism;
import Zoo.World;

public class Antelope extends Animal {
    private static final int AntelopeDodgeChance = 2;

    public Antelope(World world) {
        super(world);
        strength = 4;
        initiative = 4;
    }

    public Antelope(World world, int x, int y) {
        super(world, x, y);
        strength = 4;
        initiative = 4;
    }

    public Antelope(World world, int strength, int initiative, int x, int y, int age) {
        super(world, strength, initiative, x, y, age);
    }

    @Override
    public Color getColor() {
        return new Color(160, 82, 45);
    }

    @Override
    protected Organism getChild(int x, int y) {
        return new Antelope(world, x, y);
    }

    @Override
    public void takeAction() {
        int[] newPos = getRandomMovePosition(x, y, 2);

        Organism organism = world.getOrganism(newPos[0], newPos[1]);

        if (organism != null) {
            organism.collision(this);
        } else {
            changePosition(newPos[0], newPos[1]);
        }
    }

    @Override
    public void collision(Organism attacker) {
        if (this.getClass() == attacker.getClass()) {
            reproduce();
            return;
        }

        if (new Random().nextInt(AntelopeDodgeChance) == 0) {
            int timer = 0;
            int[] newPos = new int[2];

            Organism organism;

            do {
                if (timer > MaxChildPositionRandomisingTries) {
                    return;
                }
                newPos[0] = x;
                newPos[1] = y;
                getRandomMovePosition(newPos[0], newPos[1]);
                organism = world.getOrganism(newPos[0], newPos[1]);
                timer++;
            } while (organism != null);

            if (organism == null) {
                changePosition(newPos[0], newPos[1]);
                return;
            }
        }

        super.collision(attacker);
    }

}

File: Zoo_C++/Zoo/Antelope.cpp
#include "Antelope.h"

Organism* Antelope::getChild( int x, int y ) const
{
	return new Antelope( world, x, y );
}

Antelope::Antelope( World& world, int x, int y ) : Animal( world, x, y )
{
	strength = 4;
	initiative = 4;
}

void Antelope::takeAction()
{
	int newX = x, newY = y;

	switch (rand() % NumberOfDirections)
	{
	case Up:
		newX = newX;
		newY = abs( newY - 2 );
		break;
	case Down:
		newX = newX;
		newY = std::min( newY + 2, world.getY() - 2 );
		break;
	case Left:
		newX = abs( newX - 2 );
		newY = newY;
		break;
	case Right:
		newX = std::min( newX + 2, world.getX() - 2 );
		newY = newY;
		break;
	case UpLeft:
		newX = abs( newX - 2 );
		newY = abs( newY - 2 );
		break;
	case UpRight:
		newX = std::min( newX + 2, world.getX() - 2 );
		newY = abs( newY - 2 );
		break;
	case DownLeft:
		newX = abs( newX - 2 );
		newY = std::min( newY + 2, world.getY() - 2 );
		break;
	case DownRight:
		newX = std::min( newX + 2, world.getX() - 2 );
		newY = std::min( newY + 2, world.getY() - 2 );
		break;
	default:
		break;
	}

	Organism* organism = world.getOrganism( newX, newY );

	if (organism)
	{
		organism->collision( this );
	}
	else
	{
		changePosition( newX, newY );
	}
}

void Antelope::collision( Organism* attacker )
{
	if (typeid(*this) == typeid(*attacker))
	{
		reproduce();

		return;
	}

	if (rand() % AntelopeDodgeChance == 0)
	{
		int newX, newY, timer = 0;

		Organism* organism;

		do
		{
			if (timer > maxChildPositionRandomisingTries)
			{
				return;
			}
			newX = x;
			newY = y;
			getRandomMovePosition( newX, newY );
			organism = world.getOrganism( newX, newY );
			timer++;
		} while (organism);

		if (!organism)
		{
			changePosition( newX, newY );
			return;
		}
	}

	Animal::collision( attacker );
}

void Antelope::print( std::ostream& os ) const
{
	os << (char)14;
}

std::string Antelope::name() const
{
	return organismName;
}
